genesis_core/node/machine/pool/driver/libvirt.py

```diff
@@ -377,34 +377,43 @@ class LibvirtPoolDriverSpec(tp.NamedTuple):
     iface_mtu: int = 1450
 
 
+class LibvirtClientSingleton:
+    _instances = {}
+
+    @classmethod
+    def get_instance(cls, uri):
+        # TODO: clean up connections by TTL or when hypervisor is deleted
+        instance = cls._instances.get(uri)
+        if not instance or not instance.isAlive():
+            instance = libvirt.open(uri)
+            if not instance:
+                raise ConnectionError(
+                    f"Failed to open libvirt connection: {uri}"
+                )
+            cls._instances[uri] = instance
+        return instance
+
+
 class LibvirtPoolDriver(base.AbstractPoolDriver):
 
     def __init__(self, pool: models.MachinePool):
         self._spec = LibvirtPoolDriverSpec(**pool.driver_spec)
         self._pool = pool
+        self._client = LibvirtClientSingleton.get_instance(
```

Comment:
@akremenetsky if we instantiate only one LibvirtPoolDriver per hypervisor and work with it every time - we may remove singleton, but looks like not a bad way to leave it as is for now. Let's discuss it. Pros to remove singleton - we'll easily cleanup connections on driver GC.